src/rust/engine/src/externs/options.rs

```diff
@@ -146,11 +161,121 @@ impl PyConfigSource {
 #[pyclass]
 struct PyOptionParser(OptionParser);
 
-type RankedVal<T> = (T, isize);
+// The pythonic value of a dict-typed option.
+type PyDictVal = HashMap<String, PyObject>;
+
+// The derivation of the option value, as a vec of (value, vec of source ranks) tuples.
+// A scalar value will always have a single source. A list/dict value may have elements
+// appended across multiple sources.
+// In ascending rank order, so the last value is the final value of the option.
+type OptionValueDerivation<T> = Vec<(T, Vec<isize>)>;
+
+// A tuple (final value, rank of final value, optional derivation of value).
+//
+// Note: The final value and its rank could be computed from the derivation (see above),
+// but the full derivation is not itself computed in normal usage.
+// We could get rid of this tuple type by representing the final value and its rank as
+// a singleton derivation (in the case where we don't otherwise need the full derivation).
+// But that would allocate two unnecessary Vecs for every option.
+type OptionValue<T> = (Option<T>, isize, Option<OptionValueDerivation<T>>);
+
+// Condense list value derivation across sources, so that it reflects merges vs. replacements
```

Comment:
This logic is ported from [src/python/pants/option/parser.py](https://github.com/pantsbuild/pants/blob/9b7da2d45e2d8b72170d534f06a2d1c74cc5f21d/src/python/pants/option/parser.py#L699), which can now go away with the legacy parser.